src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp

```diff
@@ -2477,6 +2483,220 @@ address StubGenerator::generate_unsafe_copy(const char *name,
+  __ align(CodeEntryAlignment);
+  StubCodeMark mark(this, "StubRoutines", name);
+  address start = __ pc();
+
+  assert(unsafe_byte_fill != nullptr, "Invalid call");
+
+  // bump this on entry, not on exit:
+  INC_COUNTER_NP(SharedRuntime::_unsafe_set_memory_ctr, rscratch1);
+
+  {
+    Label L_exit, L_fillQuadwords, L_fillDwords, L_fillBytes;
+
+    // setup_arg_regs only has an effect for Windows
+    // Windows argument registers are moved into Linux ABI-defined registers
+    // r9 and r10 are used to save rsi and rdi and are therefore unavailable
+    // as scratch registers for the routine.
+    setup_arg_regs(3);
+
+    const Register dest = rdi;
+    const Register size = rsi;
+    const Register wide_value = rax;
+    const Register rScratch1 = rcx;
+
+    //     fill_to_memory_atomic(unsigned char*, unsigned long, unsigned char)
+
+    __ testq(size, size);
+    __ jcc(Assembler::zero, L_exit);
+
+    {
+      const Register byteVal = rdx;
+
+      // Propagate byte to full Register
+      __ movzbl(rScratch1, byteVal);
+      __ mov64(wide_value, 0x0101010101010101);
+      __ imulq(wide_value, rScratch1);
+
+      // Check for pointer & size alignment
+      __ movq(rScratch1, dest);
+      __ orq(rScratch1, size);
+    }
+
+    __ testb(rScratch1, 7);
+    __ jcc(Assembler::equal, L_fillQuadwords);
+
+    __ testb(rScratch1, 3);
+    __ jcc(Assembler::equal, L_fillDwords);
+
+    __ testb(rScratch1, 1);
+    __ jcc(Assembler::notEqual, L_fillBytes);
+
+    // Fill words
+    {
+      Label L_wordsTail, L_wordsLoop, L_wordsTailLoop;
+      UnsafeMemoryMark usmm(this, true, true);
+
+      // At this point, we know the lower bit of size is zero and a
+      // multiple of 2
+      do_setmemory_atomic_loop(USM_SHORT, dest, size, wide_value, rScratch1,
+                               L_exit, _masm);
+    }
+    __ jmp(L_exit);
+
+    __ BIND(L_fillQuadwords);
+
+    // Fill QUADWORDs
+    {
+      Label L_qwordLoop, L_qwordsTail, L_qwordsTailLoop;
+      UnsafeMemoryMark usmm(this, true, true);
+
+      // At this point, we know the lower 3 bits of size are zero and a
+      // multiple of 8
+      do_setmemory_atomic_loop(USM_QUADWORD, dest, size, wide_value, rScratch1,
+                               L_exit, _masm);
+    }
+    __ BIND(L_exit);
+
+    restore_arg_regs();
+    __ ret(0);
+
+    __ BIND(L_fillDwords);
+
+    // Fill DWORDs
+    {
+      Label L_dwordLoop, L_dwordsTail, L_dwordsTailLoop;
+      UnsafeMemoryMark usmm(this, true, true);
+
+      // At this point, we know the lower 2 bits of size are zero and a
+      // multiple of 4
+      do_setmemory_atomic_loop(USM_DWORD, dest, size, wide_value, rScratch1,
+                               L_exit, _masm);
+    }
+    __ jmp(L_exit);
+
+    __ BIND(L_fillBytes);
+    {
+      const Register byteVal = rdx;
+#ifdef _WIN32
+      __ movq(rcx, rdi); // Restore c_rarg*
+      __ movq(r8, rdx);
+      __ movq(rdx, rsi);
+      restore_arg_regs();
+#endif
```

Comment:
This is stubGenerator_x86_64.cpp 64bit specific, so WIN32 portion could be removed?

Reply:
`_WIN32` is also defined for 64 bit Windows

Reply:
Thanks, I missed that.

Reply:
I'm changing the scheme here to c_rargX, so no need for any windows-specific stuff.  I added it when I needed 4 scratch registers.